### PR TITLE
Generalise ionising photon calculation to arrays

### PR DESCRIPTION
**Pull request by stephenmwilkins**
At the moment Sed.calculate_ionising_photon_luminosity() doesn't work for arrays of spectra. This fixes this.

## Issue Type
- [x] Bug

## Checklist
- [x] I have read the [CONTRIBUTING.md]() -->
- [x] I have added docstrings to all methods
- [x] I have added sufficient comments to all lines
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no pep8 errors
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/synthesizer/sed.py [hunk 1]
```diff
@@ -1314,18 +1314,25 @@ def calculate_ionising_photon_production_rate(
...
-            ionisation_wavelength.to(angstrom).value, x, y
-        )
-
         # Restrict arrays to ionisation regime
         x = x[ionisation_mask]
-        y = y[ionisation_mask]
+        if len(y.shape) == 1:
+            y = y[ionisation_mask]
+        else:
+            y = y[:, ionisation_mask]
```
Comment (WillJRoper) [inline src/synthesizer/sed.py:1322]: This only works for 1D and 2D now, what about 3D?  

Use
```
masked_array = arr[..., mask]
```
to mask the final axis of an arbitrarily dimensional array.

File: src/synthesizer/sed.py [hunk 2]
```diff
@@ -1314,18 +1314,25 @@ def calculate_ionising_photon_production_rate(
...
+
+        # Add a final data point at the ionising energy to ensure full
+        # coverage.
+        x0 = ionisation_wavelength.to(angstrom).value
+        if len(y.shape) == 1:
```
Comment (WillJRoper) [inline src/synthesizer/sed.py:1327]: Same with this block, it'll fail for 3D arrays. You could use scipy and `interp1d` with the axis argument or I think you could just use pure numpy with `apply_along_axis`
```
y0 = np.apply_along_axis(lambda y_: np.interp(x0, x, y_), axis=-1, arr=y)
``` 
https://numpy.org/doc/2.0/reference/generated/numpy.apply_along_axis.html

**Review by WillJRoper — CHANGES_REQUESTED** [review]

File: src/synthesizer/sed.py (hunk 1 above)
Comment (stephenmwilkins) [reply src/synthesizer/sed.py]: Fixed.

File: src/synthesizer/sed.py (hunk 2 above)
Comment (stephenmwilkins) [reply src/synthesizer/sed.py:1327]: Fixed. See comment below.

**Comment by stephenmwilkins** [comment]
I've generalised it nD and added some examples to the docs demonstrating this. However, this revealed that several other methods on Sed don't work in 3D. I'll raise an issue for these.

File: docs/source/sed/sed_example.ipynb [hunk 3]
```diff
@@ -509,13 +509,47 @@
...
+   "execution_count": null,
+   "id": "0cbf660f",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sed3 = grid.get_sed(spectra_type='incident')\n",
+    "print(sed3.lnu.shape)\n",
+    "\n",
+    "print(sed3.measure_window_lnu((1400, 1600) * Angstrom))\n",
+    "# print(sed3.measure_beta())  # DOES NOT WORK\n",
```
Comment (WillJRoper) [inline docs/source/sed/sed_example.ipynb:544]: Can't merge stuff like this into main for it to appear in the docs. Please remove it or fix it.

**Review by WillJRoper — CHANGES_REQUESTED** [review]

File: docs/source/sed/sed_example.ipynb (hunk 3 above)
Comment (stephenmwilkins) [reply docs/source/sed/sed_example.ipynb]: Done

**Review by WillJRoper — APPROVED** [review]

**Pull request merged**